File: swift_calculator/calculator/calculator/CalculatorModel.swift
//
//  CalculatorModel.swift
//  calculator
//
//  Created by JOSE ZARABANDA on 7/25/25.
//

import Foundation

class CalculatorModel: ObservableObject {
    @Published var display: String = "0"
    
    private var currentValue: Double = 0
    private var previousValue: Double = 0
    private var currentOperation: Operation?
    private var shouldResetDisplay: Bool = false
    
    enum Operation: String, CaseIterable {
        case add = "+"
        case subtract = "−"
        case multiply = "×"
        case divide = "÷"
        
        var symbol: String {
            return self.rawValue
        }
        
        func calculate(_ a: Double, _ b: Double) -> Double {
            switch self {
            case .add:
                return a + b
            case .subtract:
                return a - b
            case .multiply:
                return a * b
            case .divide:
                guard b != 0 else { return Double.nan }
                return a / b
            }
        }
    }
    
    func numberPressed(_ number: Int) {
        if shouldResetDisplay {
            display = "0"
            shouldResetDisplay = false
        }
        
        if display == "0" {
            display = String(number)
        } else {
            display += String(number)
        }
        
        currentValue = Double(display) ?? 0
    }
    
    func decimalPressed() {
        if shouldResetDisplay {
            display = "0"
            shouldResetDisplay = false
        }
        
        if !display.contains(".") {
            display += "."
        }
    }
    
    func operationPressed(_ operation: Operation) {
        if currentOperation != nil && !shouldResetDisplay {
            calculateResult()
        }
        
        previousValue = currentValue
        currentOperation = operation
        shouldResetDisplay = true
    }
    
    func equalsPressed() {
        calculateResult()
        currentOperation = nil
        shouldResetDisplay = true
    }
    
    func clearPressed() {
        display = "0"
        currentValue = 0
        previousValue = 0
        currentOperation = nil
        shouldResetDisplay = false
    }
    
    func clearEntryPressed() {
        display = "0"
        currentValue = 0
        shouldResetDisplay = false
    }
    
    func toggleSignPressed() {
        if display == "0" || display == "Error" || display == "∞" {
            return
        }
        
        if display.hasPrefix("-") {
            display = String(display.dropFirst())
        } else {
            display = "-" + display
        }
        
        currentValue = Double(display) ?? 0
    }
    
    private func calculateResult() {
        guard let operation = currentOperation else { return }
        
        let result = operation.calculate(previousValue, currentValue)
        
        if result.isNaN {
            display = "Error"
            currentValue = 0
        } else if result.isInfinite {
            display = "∞"
            currentValue = 0
        } else {
            currentValue = result
            display = formatResult(result)
        }
    }
    
    private func formatResult(_ value: Double) -> String {
        // Handle special cases
        if value.isNaN {
            return "Error"
        }
        if value.isInfinite {
            return value > 0 ? "∞" : "-∞"
        }
        
        // Check if the number is an integer
        if value.truncatingRemainder(dividingBy: 1) == 0 && abs(value) < 1e15 {
            return String(format: "%.0f", value)
        }
        
        // Format decimal numbers
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 8
        formatter.minimumFractionDigits = 0
        formatter.usesGroupingSeparator = false
        
        if let formatted = formatter.string(from: NSNumber(value: value)) {
            return formatted
        }
        
        // Fallback for very large/small numbers
        if abs(value) >= 1e15 || (abs(value) < 1e-6 && value != 0) {
            return String(format: "%.2e", value)
        }
        
        return String(value)
    }
}

File: fluter_calculator/lib/calculator_model.dart
import 'package:flutter/foundation.dart';

enum Operation {
  add,
  subtract,
  multiply,
  divide,
}

extension OperationExtension on Operation {
  String get symbol {
    switch (this) {
      case Operation.add:
        return '+';
      case Operation.subtract:
        return '−';
      case Operation.multiply:
        return '×';
      case Operation.divide:
        return '÷';
    }
  }

  double calculate(double a, double b) {
    switch (this) {
      case Operation.add:
        return a + b;
      case Operation.subtract:
        return a - b;
      case Operation.multiply:
        return a * b;
      case Operation.divide:
        if (b == 0) return double.nan;
        return a / b;
    }
  }
}

class CalculatorModel extends ChangeNotifier {
  String _display = "0";
  double _currentValue = 0;
  double _previousValue = 0;
  Operation? _currentOperation;
  bool _shouldResetDisplay = false;

  String get display => _display;

  void numberPressed(int number) {
    if (_shouldResetDisplay) {
      _display = "0";
      _shouldResetDisplay = false;
    }

    if (_display == "0") {
      _display = number.toString();
    } else {
      _display += number.toString();
    }

    _currentValue = double.tryParse(_display) ?? 0;
    notifyListeners();
  }

  void decimalPressed() {
    if (_shouldResetDisplay) {
      _display = "0";
      _shouldResetDisplay = false;
    }

    if (!_display.contains(".")) {
      _display += ".";
    }
    notifyListeners();
  }

  void operationPressed(Operation operation) {
    if (_currentOperation != null && !_shouldResetDisplay) {
      _calculateResult();
    }

    _previousValue = _currentValue;
    _currentOperation = operation;
    _shouldResetDisplay = true;
    notifyListeners();
  }

  void equalsPressed() {
    _calculateResult();
    _currentOperation = null;
    _shouldResetDisplay = true;
    notifyListeners();
  }

  void clearPressed() {
    _display = "0";
    _currentValue = 0;
    _previousValue = 0;
    _currentOperation = null;
    _shouldResetDisplay = false;
    notifyListeners();
  }

  void clearEntryPressed() {
    _display = "0";
    _currentValue = 0;
    _shouldResetDisplay = false;
    notifyListeners();
  }

  void toggleSignPressed() {
    if (_display == "0" || _display == "Error" || _display == "∞") {
      return;
    }

    if (_display.startsWith("-")) {
      _display = _display.substring(1);
    } else {
      _display = "-$_display";
    }

    _currentValue = double.tryParse(_display) ?? 0;
    notifyListeners();
  }

  void _calculateResult() {
    if (_currentOperation == null) return;

    final result = _currentOperation!.calculate(_previousValue, _currentValue);

    if (result.isNaN) {
      _display = "Error";
      _currentValue = 0;
    } else if (result.isInfinite) {
      _display = "∞";
      _currentValue = 0;
    } else {
      _currentValue = result;
      _display = _formatResult(result);
    }
  }

  String _formatResult(double value) {
    // Handle special cases
    if (value.isNaN) {
      return "Error";
    }
    if (value.isInfinite) {
      return value > 0 ? "∞" : "-∞";
    }

    // Check if the number is an integer
    if (value % 1 == 0 && value.abs() < 1e15) {
      return value.toInt().toString();
    }

    // Format decimal numbers
    String formatted = value.toString();
    
    // Remove trailing zeros after decimal point
    if (formatted.contains('.')) {
      formatted = formatted.replaceAll(RegExp(r'0*$'), '');
      formatted = formatted.replaceAll(RegExp(r'\.$'), '');
    }

    // Fallback for very large/small numbers
    if (value.abs() >= 1e15 || (value.abs() < 1e-6 && value != 0)) {
      return value.toStringAsExponential(2);
    }

    return formatted;
  }
}
